proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+
+Since this proposal is completely backwards compatible, GADT-syntax
+``newtype``-s must behave differently without
+``-XLinearTypes``. GADT-syntax ``data`` definitions need not, but it
+is the expectation of the programmer that the following two are
+equivalent definitions (which they are not with ``-XLinearTypes``):
+
+::
+
+  data Maybe a
+    = Just a
+    | Nothing
+
+  data Maybe a where
+    Just :: a -> Maybe a
+    Nothing :: Maybe a
+
+To follow the principle of least surprise (which we take to mean that
+only programmers aware of ``-XLinearTypes`` would be surprised), we
+interpret GADT-syntax type declaration (both ``data`` and ``newtype``)
+in code without ``-XLinearTypes`` to be *linear*, despite the
+ostensible use of an unrestricted arrow.
+
+Costs and Drawbacks
+===================
+
+Learnability
+------------
+
+This proposal tries hard to make the changes unintrusive to newcomers,
+or indeed to the existing language ecosystem as a whole. However, if
+many users start adopting it, inevitably, linear arrows may start
+appearing in so many libraries that it becomes hard to be oblivious to
+their existence. They can be safely ignored, but teachers of Haskell
+might still consider them distracting for their students.
+
+Development and maintenance
+---------------------------
+
+The arrow type constructor is constructed and destructed a lot in
+GHC's internals. So there are many places in the type checker where
+the GHC implementation will have to handle multiplicities. It is most
+often straightforward as it consists in getting a multiplicity
+variable and pass it to a function. Nevertheless, it is possible to
+get it wrong. And type checker developers will have to be aware of
+multiplicities to modify most aspects of type checking.
+
+Linear types also affect Core: Core must handle linear types, and the
+linter modified accordingly to check linearity, in order to ensure
+that core-to-core passes do not break the linearity guarantees. The
+flip side is that all core-to-core passes must make sure that they do
+not break linearity. It is possible that some of the pre-linear-type
+passes actually do break linearity in some cases (note: there has been
+no evidence of this so far).
+
+Unification of multiplicity expressions (as for instance in the type
+of ``(.)`` above) requires some flavour of unification module
+associativity and commutativity (AC). Unification modulo AC is
+well-understood an relatively easy to implement. But would still be
+a non-trivial addition to the type-checker. We may decide that
+a simplified fragment is better suited for our use-case that the full
+generality of AC.
+
+Alternatives
+============
+
+.. _Alternatives:
+
+This section describes variants that could be considered for inclusion
+in the proposal.
+
+Lexical tokens of the multiplicity-parametric arrow
+---------------------------------------------------
+
+Here are the potential notations for the multiplicity-parametric
```

Comment:
31. I still like `|p->`, just because it doesn't steal syntax; I do think it should be listed as an alternative. The "line" comment was about how the syntax can be a mnemonic for the idea of *line*ar types.